apps/prairielearn/src/ee/lib/ai-grading.sql

```diff
@@ -24,3 +35,61 @@ ORDER BY
   s.date DESC
 LIMIT
   1;
+
+-- BLOCK select_embedding_for_submission
+SELECT
+  *
+FROM
+  submission_grading_context_embeddings AS emb
+WHERE
+  emb.submission_id = $submission_id;
+
+-- BLOCK update_embedding_for_submission
+INSERT INTO
+  submission_grading_context_embeddings (embedding, submission_id, submission_text)
+VALUES
+  ($embedding, $submission_id, $submission_text);
+
+-- BLOCK select_closest_embeddings
```

Comment:
Have you looked at what the query planner will do with this? Here's an example run against our production database with the embeddings stuff omitted:

```
                                                                      QUERY PLAN
-------------------------------------------------------------------------------------------------------------------------------------------------------
 Nested Loop  (cost=27603537.52..27603551.32 rows=1 width=16)
   ->  Nested Loop  (cost=27603536.95..27603547.85 rows=1 width=8)
         ->  Nested Loop  (cost=27603536.38..27603544.43 rows=1 width=8)
               ->  GroupAggregate  (cost=27603535.82..27603535.83 rows=1 width=16)
                     Group Key: submissions.id
                     ->  Sort  (cost=27603535.82..27603535.82 rows=1 width=8)
                           Sort Key: submissions.id
                           ->  Gather  (cost=19722086.45..27603535.81 rows=1 width=8)
                                 Workers Planned: 2
                                 ->  Parallel Hash Join  (cost=19721086.45..27602535.71 rows=1 width=8)
                                       Hash Cond: ((grading_jobs.submission_id = submissions.id) AND (grading_jobs.graded_at = submissions.graded_at))
                                       ->  Parallel Seq Scan on grading_jobs  (cost=0.00..6904736.30 rows=48975230 width=16)
                                       ->  Parallel Hash  (cost=18738437.21..18738437.21 rows=49422016 width=16)
                                             ->  Parallel Seq Scan on submissions  (cost=0.00..18738437.21 rows=49422016 width=16)
                                                   Filter: (id <> 92534478)
               ->  Index Scan using submissions_pkey on submissions s  (cost=0.57..8.59 rows=1 width=16)
                     Index Cond: (id = submissions.id)
         ->  Index Scan using variants_pkey on variants v  (cost=0.57..3.42 rows=1 width=16)
               Index Cond: (id = s.variant_id)
   ->  Index Scan using instance_questions_pkey on instance_questions iq  (cost=0.57..3.46 rows=1 width=16)
         Index Cond: (id = v.instance_question_id)
         Filter: (assessment_question_id = 26489681)
(22 rows)
```

This does sequential scans on both `submissions` and `grading_jobs`. This will have unacceptably slow performance in production. Please rewrite this to be more performant.

Reply:
```
QUERY PLAN                                       
                                               
---------------------------------------------------------------------------------------------------------------------------------------
-----------------------------------------------
 Nested Loop  (cost=21.49..23.08 rows=1 width=48)
   Join Filter: (ls.s_id = emb.submission_id)
   ->  Subquery Scan on ls  (cost=21.49..21.88 rows=1 width=24)
         Filter: (ls.rn = 1)
         ->  WindowAgg  (cost=21.49..21.73 rows=12 width=48)
               Run Condition: (row_number() OVER (?) <= 1)
               ->  Sort  (cost=21.49..21.52 rows=12 width=40)
                     Sort Key: s.variant_id, s.date DESC
                     ->  Nested Loop  (cost=2.71..21.28 rows=12 width=40)
                           ->  Hash Join  (cost=2.56..15.98 rows=15 width=24)
                                 Hash Cond: (v.instance_question_id = iq.id)
                                 ->  Seq Scan on variants v  (cost=0.00..12.70 rows=270 width=16)
                                 ->  Hash  (cost=2.54..2.54 rows=2 width=16)
                                       ->  Seq Scan on instance_questions iq  (cost=0.00..2.54 rows=2 width=16)
                                             Filter: ((NOT requires_manual_grading) AND (status <> 'unanswered'::enum_instance_question
_status) AND (assessment_question_id = 18691))
                           ->  Index Scan using submissions_variant_id_idx on submissions s  (cost=0.14..0.34 rows=1 width=24)
                                 Index Cond: (variant_id = v.id)
                                 Filter: (id <> 4)
   ->  Seq Scan on submission_grading_context_embeddings emb  (cost=0.00..1.09 rows=9 width=40)
(19 rows)
```

I just wrote a new version of the function that's generating this query plan. Is it more acceptable now?

Reply:
This is much, much better, yeah. For some sample data in production (of course without the embeddings table since that doesn't exist yet), the total time is now ~2ms:

```
 Subquery Scan on ls  (cost=60551.19..60579.57 rows=4 width=16) (actual time=0.980..1.033 rows=50 loops=1)
   Filter: (ls.rn = 1)
   ->  WindowAgg  (cost=60551.19..60568.65 rows=873 width=48) (actual time=0.979..1.025 rows=50 loops=1)
         Run Condition: (row_number() OVER (?) <= 1)
         ->  Sort  (cost=60551.19..60553.38 rows=873 width=32) (actual time=0.969..0.973 rows=61 loops=1)
               Sort Key: s.variant_id, s.date DESC
               Sort Method: quicksort  Memory: 29kB
               ->  Nested Loop  (cost=1.70..60508.55 rows=873 width=32) (actual time=0.082..0.912 rows=61 loops=1)
                     ->  Nested Loop  (cost=1.14..42065.11 rows=617 width=24) (actual time=0.068..0.559 rows=60 loops=1)
                           ->  Index Scan using instance_questions_assessment_question_id_assessment_instan_key on instance_questions iq  (cost=0.57..2077.37 rows=523 width=16) (actual time=0.054..0.266 rows=29 loops=1)
                                 Index Cond: (assessment_question_id = 26489681)
                                 Filter: ((NOT requires_manual_grading) AND (status <> 'unanswered'::enum_instance_question_status))
                                 Rows Removed by Filter: 119
                           ->  Index Scan using variants_instance_question_id_number_key on variants v  (cost=0.57..76.24 rows=22 width=16) (actual time=0.008..0.009 rows=2 loops=29)
                                 Index Cond: (instance_question_id = iq.id)
                     ->  Index Scan using submissions_variant_id_idx on submissions s  (cost=0.57..29.59 rows=30 width=16) (actual time=0.005..0.005 rows=1 loops=60)
                           Index Cond: (variant_id = v.id)
                           Filter: (id <> 1)
 Planning Time: 0.837 ms
 Execution Time: 1.331 ms
(20 rows)
```